cstar/scheduler/job.py

```diff
@@ -0,0 +1,1001 @@
+
+        # Explicitly typing to avoid mypy confusion in conditional pathways below
+        self.cpus_per_node: Optional[int]
+        self.nodes: Optional[int]
+
+        if (
+            (nodes is None)
+            and (cpus_per_node is not None)
+            and (scheduler.requires_task_distribution)
+        ):
+            self.nodes = ceil(cpus / cpus_per_node)
+            self.cpus_per_node = cpus_per_node
+        elif (
+            (nodes is not None)
+            and (cpus_per_node is None)
+            and (scheduler.requires_task_distribution)
+        ):
+            self.nodes = nodes
+            self.cpus_per_node = int(cpus / nodes)
+        elif (
+            (nodes is None)
+            and (cpus_per_node is None)
+            and (scheduler.requires_task_distribution)
+        ):
+            if scheduler.global_max_cpus_per_node is None:
+                raise EnvironmentError(
+                    "You attempted to create a scheduler job without 'nodes', and "
+                    + "'cpus_per_node' parameters, but your scheduler explicitly "
+                    + "requires a task distribution. C-Star is unable to determine "
+                    + "your system's CPUs per node automatically and cannot continue"
+                )
+
+            nnodes, ncpus = self._calculate_node_distribution(
+                cpus, scheduler.global_max_cpus_per_node
+            )
+            warnings.warn(
+                (
+                    "WARNING: Attempting to create scheduler job without 'nodes' and 'cpus_per_node' "
+                    + "parameters, but your system requires an explicitly specified task distribution."
+                    + "\n C-Star will attempt "
+                    + f"\nto use a distribution of {nnodes} nodes with {ncpus} CPUs each, "
+                    + "\nbased on your system maximum of "
+                    + f"{scheduler.global_max_cpus_per_node} CPUS per node "
+                    + f"\nand your job requirement of {cpus} CPUS."
+                ),
+                UserWarning,
+            )
+            self.cpus_per_node = ncpus
+            self.nodes = nnodes
+        else:
+            self.cpus_per_node = cpus_per_node
+            self.nodes = nodes
+
+        self.account_key = account_key
+        self._id: Optional[int] = None
+
+    @property
+    def id(self) -> Optional[int]:
+        """Retrieve the unique job ID assigned by the scheduler.
+
+        The job ID is assigned when the job is successfully submitted to the scheduler.
+        If the job has not been submitted, a message will be displayed indicating that
+        the ID is not yet available.
+
+        Returns
+        -------
+        id: int or None
+            The unique job ID assigned by the scheduler, or `None` if the job has not
+            been submitted.
+        """
+
+        if self._id is None:
+            print("No Job ID found. Submit this job with SchedulerJob.submit()")
```

Comment:
Could also set it as `0` or something as a placeholder if using a fixed type is particularly useful. I think a raise is a bit aggressive here, it's valid to query the id of a job that doesn't have one yet and I'd expect a gentle reminder rather than an error.